### DISCUSSION TITLE
Error occurred when added multiple plotters into a qt window

### DISCUSSION THREAD

**Discussion by matudinglin** [pyvista/pyvista#4564] [Q&A]
Hi there, I'm trying to implement multiple rendering window into qt window. 
This is my code: 
```
self.plotter = QtInteractor(self.ui.plotter_splitter)
self.plotter_2 = QtInteractor(self.ui.plotter_splitter)
```
After exiting the app, errors will occur: 

> 2023-06-22 16:21:11.565 (  23.784s) [                ]vtkWin32OpenGLRenderWin:267    ERR| vtkWin32OpenGLRenderWindow (000001F186AC4910): wglMakeCurrent failed in MakeCurrent(), error: Ω¢░Φù¼╟▒

Could you solve these problems for me? Is there a better way to manage multiple plotter windows within a qt window? Thanks.

**Comment by banesullivan** [comment]
cc @pyvista/pyvistaqt-admin

**Comment by larsoner** [comment]
In MNE we use `plotter.subplot` and it seems to work. I don't think we've tried embedding more than one QtInteractor

**Reply by matudinglin** [reply]
Hi Eric, thanks for your reply. if It's not possible to embed multiple QtInteractor. Is there a way to add subplot at runtime, so that the user can add more rendering window as they wish?

**Reply by larsoner** [reply]
Have you tried using `.subplot` / Plotter methods to add/remove subplots? In theory they might work at runtime but I haven't checked

In principle it should also be possible to use multiple QtInteractors, too, I think it's probably a bug that it doesn't work

**Reply by matudinglin** [reply]
I looked at the code and it seems that `.subplot` method can only be used to set current subplot windows. 

Maybe I implemented multiple QtInteractors in the wrong way? But I have no idea how to fix this.

**Reply by sylvanoMTL** [reply]
I think this is related to vtk. There is this post: https://stackoverflow.com/questions/73457615/pyqt5-vtk-wglmakecurrent-failed-error, where the solution is to close the widget, which would call the function Finalize(). 

pyvistaqt examples does not provide enough indication on how to closes plots or renderer.

I guess that when closing the main window a vtk renderer is still active and throw these vtkWin32OpenGLRenderWindow error, as the window/widget has been closed, but not the render. 

I am testing to add a `def closeEvent(self, QCloseEvent):`  but trying to close the plotter, or getting the renderer and closing it have been unsuccessful

**Reply by moonlitfjords** [reply]
> I think this is related to vtk. There is this post: https://stackoverflow.com/questions/73457615/pyqt5-vtk-wglmakecurrent-failed-error, where the solution is to close the widget, which would call the function Finalize().
> 
> pyvistaqt examples does not provide enough indication on how to closes plots or renderer.
> 
> I guess that when closing the main window a vtk renderer is still active and throw these vtkWin32OpenGLRenderWindow error, as the window/widget has been closed, but not the render.
> 
> I am testing to add a `def closeEvent(self, QCloseEvent):` but trying to close the plotter, or getting the renderer and closing it have been unsuccessful

I had been having this issue as well, and just sort of ignored it after seeing this post in the hope that when I came back to it later you or someone else would have found a fix.

I just tried taking another look at it though and seem to have managed to resolve it, at least in my scenario, so thought I would share in case it is also applicable to others (I previously spent a long time going through many such posts as this without finding a solution). I don't know if this is due to something that has changed within the source code, as I thought I had tried this before to no avail, but it is also possible that I just didn't try this particular combination for whatever reason.

Anyway, I found that if you store each `QtInteractor` instance (e.g. as `self.plotter1 = QtInteractor()` for instance) so you can refer back to it later, and override the application's `closeEvent` as @sylvanoMTL was suggesting, then calling `plotter.Finalize()` followed by `plotter.close()` for each of these instances in turn seems to have done the job for me. I haven't tested this extensively, but it is the first time in many months that I haven't seen a stream of red whenever I close my application, so if it is of use to other people, either in the short-term or in implementing a proper fix (or documentation of this usage) then that would be great.
